File: dailyQuestion/2020/2020-08/08-02/python/solution_preorder.py

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def flatten(root: TreeNode) -> None:
    """
    Do not return anything, modify root in-place instead.
    """
    if not root:
        return
    stack = [root]
    pre = None
    while stack:
        node = stack.pop()
        if pre:
            pre.left = None
            pre.right = node
        if node.right:
            stack.append(node.right)
        if node.left:
            stack.append(node.left)
        pre = node

if __name__ == "__main__" :
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(5)
    root.left.left = TreeNode(3)
    root.left.right = TreeNode(4)
    root.right.right = TreeNode(6)
    flatten(root)

File: dailyQuestion/2020/2020-08/08-02/golang/solution_preorder.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func flatten(root *TreeNode) {
	list := preorder(root)
	for i := 1; i < len(list); i++ {
		pre, cur := list[i-1], list[i]
		pre.Left = nil
		pre.Right = cur
	}
}

func preorder(root *TreeNode) []*TreeNode {
	result := []*TreeNode{}
	if root == nil {
		return nil
	}
	result = append(result, root)
	result = append(result, preorder(root.Left)...)
	result = append(result, preorder(root.Right)...)
	return result
}
func main() {
	root := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 5}}
	root.Left.Left = &TreeNode{Val: 3}
	root.Left.Right = &TreeNode{Val: 4}
	root.Right.Right = &TreeNode{Val: 6}
	flatten(root)
}
